singularity-eos/eos/eos_vinet.hpp

```diff
@@ -0,0 +1,374 @@
+  // overloads that use the scalar versions declared here
+  SG_ADD_BASE_CLASS_USINGS(Vinet)
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return 0; }
+  static constexpr unsigned long PreferredInput() { return _preferred_input; }
+  PORTABLE_INLINE_FUNCTION void PrintParams() const {
+    static constexpr char st[]{"Vinet Params: "};
+    printf("%s rho0:%e T0:%e B0:%e BP0:%e\n  A0:%e Cv0:%e E0:%e S0:%e\n", st, _rho0, _T0,
+           _B0, _BP0, _A0, _Cv0, _E0, _S0);
+    for (int i = 0; i < 39; i++) {
+      if (_d2to40[i] > 0.0) printf("d%i:%e\t", i + 2, _d2to40[i]);
+    }
+    printf("\n\n");
+  }
+  // Density/Energy from P/T not unique, if used will give error
+  PORTABLE_INLINE_FUNCTION void
+  DensityEnergyFromPressureTemperature(const Real press, const Real temp, Real *lambda,
+                                       Real &rho, Real &sie) const;
+  inline void Finalize() {}
+  static std::string EosType() { return std::string("Vinet"); }
+  static std::string EosPyType() { return EosType(); }
+
+ private:
+  // AEM: I am not sure what this does so leave it in.
+  static constexpr const unsigned long _preferred_input =
+      thermalqs::density | thermalqs::temperature;
+  Real _rho0, _T0, _B0, _BP0, _A0, _Cv0, _E0, _S0;
+  const Real *_expconsts;
+  Real _VIP[43], _d2to40[39];
+  PORTABLE_INLINE_FUNCTION void CheckVinet();
+  PORTABLE_INLINE_FUNCTION void Vinet_F_DT_func(const Real rho, const Real T,
+                                                Real *output) const;
+};
+
+PORTABLE_INLINE_FUNCTION void Vinet::CheckVinet() {
+
+  int ind;
+
+  if (_rho0 < 0.0) {
+    printf("testing rho0\n");
+    EOS_ERROR("Required Vinet model parameter rho0 < 0");
+    return;
+  }
+  if (_T0 < 0.0) {
+    printf("testing T0\n");
+    EOS_ERROR("Required Vinet model parameter T0 < 0");
+    return;
+  }
+  if (_B0 < 0.0) {
+    printf("testing B0\n");
+    EOS_ERROR("Required Vinet model parameter B0 < 0");
+    return;
+  }
+  if (_BP0 < 1.0) {
+    printf("testing BP0\n");
+    EOS_ERROR("Required Vinet model parameter BP0 < 1");
+    return;
+  }
+  if (_A0 < 0.0) {
+    printf("testing A0\n");
+    EOS_ERROR("Required Vinet model parameter A0 < 0");
+    return;
+  }
+  if (_Cv0 < 0.0) {
+    printf("testing Cv0\n");
+    EOS_ERROR("Required Vinet model parameter Cv0 < 0");
+    return;
+  }
+
+  for (ind = 0; ind < 39; ind++)
+    _d2to40[ind] = _expconsts[ind];
+
+  _VIP[0] = _B0 / _rho0 + pow(_A0 * _B0 / _rho0, 2.0) * _T0 / _Cv0; // sound speed squared
+  _VIP[1] = 3.0 / 2.0 * (_BP0 - 1.0);                               // exponent
+  _VIP[2] = 1.0;                                                    // prefactor f0
+  _VIP[3] = 0.0;                                                    // prefactor f1
+
+  for (ind = 4; ind < 43; ind++)
+    _VIP[ind] = _d2to40[ind - 4];
+  ind = 42;
+  while (_VIP[ind] == 0.0)
+    ind--;
+  while (ind >= 4) {
+    ind--;
+    _VIP[ind] = _VIP[ind] - (ind - 1) / _VIP[1] * _VIP[ind + 1];
+  }
```

Comment:
A few things here:
1. Is there a divide by zero possibility for the first iteration of the final loop because _VIP[ind + 1]` is zero by definition? Is there also the possibility of out of bounds memory access if `_VIP[ind]` is nonzero for every element?
2. As a general rule, we try to avoid overflow/underflow in division, which can happen when dividing by small numbers. Instead of `a / b` use `robust::ratio(a, b)` which will safely prevent division by zero assuming that neither a or b is *actually* zero. So point 1 above is still relevant.
3. Where does that formula `_VIP[ind] = _VIP[ind] - (ind - 1) / _VIP[1] * _VIP[ind + 1];` come from? Is it an integral? A description would be useful.
4. It's bad code smell to use loops that don't have obvious endpoints or easy to parse flow control. So it's better to map these `while` loops into `for` loops. For example, if the size of the `expconsts` array is known, which it will be because we'll pass it in somehow as discussed above, you don't need to do the middle while loop. The first loop can also be avoided with careful tracking of indices.
```C++
for (int i = expconsts.size() - 1; i >= 4; i--) {
  _VIP[i] = _d2to40[i-4] - (i - 1) / _VIP[1] * _VIP[ind + 1];
}
```
which to my mind is more legible and clear (assuming I transformed it correctly). I didn't fix the possible out of bounds issue at the edge of the array though. A safer version is probably:
```C++
// boundary condition
_VIP[exponcsts.size() - 1 + 4] = (expconsts.size() == 43) ? 0 : _d2to40[exponcsts.size() - 1] ;
// main loop
for (int i = expconsts.size() - 2 + 4; i >= 4; i--) {
  _VIP[i] = _d2to40[i-4] - robust::ratio(i - 1,  _VIP[1]) * _VIP[ind + 1];
}
```
or something like that.

Reply:
I see @jhp-lanl made a very similar comment after I wrote this. Oops. Well I agree with Jeff. ;)